[EVM-Equivalence-YUL] Remove calls to mstore8

vladimirradosavljevic opened the pull request:
# What ❔
This removes calls to mstore8, since they are expensive (this call will be translated to at least 4 instructions).
<!-- What are the changes this PR brings about? -->
<!-- Example: This PR adds a PR template to the repo. -->
<!-- (For bigger PRs adding more context is appreciated) -->

## Why ❔

<!-- Why are these changes done? What goal do they contribute to? What are the principles behind them? -->
<!-- Example: PR templates ensure PR reviewers, observers, and future iterators are in context about the evolution of repos. -->

## Checklist

<!-- Check your PR fulfills the following items. -->
<!-- For draft PRs check the boxes as you complete them. -->

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.

vladimirradosavljevic commented on system-contracts/contracts/EvmInterpreterFunctions.template.yul:
```diff
@@ -310,11 +301,7 @@ function getDeployedBytecode() {
-    mstore8(1, 0xc1)
-    mstore8(2, 0x4e)
-    mstore8(3, 0x9e)
+    mstore(0, 0x04C14E9E00000000000000000000000000000000000000000000000000000000)
```
@jrchatruc Is there a possibility that this will override some bytes that are important before the call?

gianbelinche replied:
There is no problem, all test pass when switching to `mstore`

vladimirradosavljevic commented on system-contracts/contracts/EvmInterpreterFunctions.template.yul:
```diff
@@ -801,13 +760,8 @@ function _pushEVMFrame(_passGas, _isStatic) {
-    mstore8(1, 0x67)
-    mstore8(2, 0xd2)
-    mstore8(3, 0xf0)
+    mstore(0, 0xE467D2F000000000000000000000000000000000000000000000000000000000)
```
@jrchatruc Is there a possibility that this will override some bytes that are important before the call?

vladimirradosavljevic commented on system-contracts/contracts/EvmInterpreterFunctions.template.yul:
```diff
@@ -1235,12 +1189,7 @@ function isAddrEmpty(addr) -> isEmpty {
-    mstore8(1, 0xe5)
-    mstore8(2, 0xab)
-    mstore8(3, 0x4a)
+    mstore(0, 0x24E5AB4A00000000000000000000000000000000000000000000000000000000)
```
@jrchatruc Is there a possibility that this will override some bytes that are important before the call?

gianbelinche replied:
There is no problem, all test pass when switching to `mstore`

vladimirradosavljevic commented on system-contracts/contracts/EvmInterpreterLoop.template.yul:
```diff
@@ -459,12 +459,7 @@ for { } true { } {
-                shr(248, mload(add(offset, i)))
-            )
-        }
+        mcopy(dst, offset, len)
```
@jrchatruc Is there a chance that dst and offset pointers can overlap? If that can never happen, we could instead of calling mcopy (essentially optimized implementation of memmove) we could implement memcpy that is efficient that this.

gianbelinche replied:
Yes, `dst` and `offset` can overlap, from `MCOPY` in https://www.evm.codes/ :

> Copying takes place as if an intermediate buffer was used, allowing the destination and source to overlap

vladimirradosavljevic replied:
Since this is change for `OP_CODECOPY`, my question was whether `dst` and `offset` can overlap in for opcode? Please note that mcopy function handles overlap case.

gianbelinche replied:
No they can't, since `offset` is from the bytecode memory space, while `dst` is from the heap memory space.
A check is being performed to make sure they are inside their respective memory space.

This is the check:

```
        // Check bytecode overflow
        if gt(add(offset, len), sub(MEM_OFFSET(), 1)) {
            revertWithGas(evmGasLeft)
        }
```

vladimirradosavljevic replied:
In that case, I have implemented memcpy function that is faster than mcopy, since mcopy handles overlap  and here, we don't need to worry about.

gianbelinche commented:
> @jrchatruc Do you recall a reason why `OP_MCOPY` is implemented with calls to `mstore8` instead of calling `mcopy`? If there was a problem with `mcopy` implementation, then I suggest we need to investigate that since `mcopy` implementation is optimized inplace memmove implementation and if there is a problem with that, we would need to find a culprit.

It used to break some tests, it does not do it anymore, so that change is fine

vladimirradosavljevic commented:
> > @jrchatruc Do you recall a reason why `OP_MCOPY` is implemented with calls to `mstore8` instead of calling `mcopy`? If there was a problem with `mcopy` implementation, then I suggest we need to investigate that since `mcopy` implementation is optimized inplace memmove implementation and if there is a problem with that, we would need to find a culprit.
> 
> It used to break some tests, it does not do it anymore, so that change is fine

Thanks for the review!
I submitted a change for 'OP_MCOPY' to use `mcopy` function instead of calling `mstore8`. AFAIU, previous implementation was first copying data into intermediate buffer, then to the destination? If this is true, do we need following check:
```
        let oldSize := mul(mload(MEM_OFFSET()),32)
        if gt(add(oldSize,size),MAX_POSSIBLE_MEM()) {
            revertWithGas(evmGasLeft)
        }
```

gianbelinche commented:
> > > @jrchatruc Do you recall a reason why `OP_MCOPY` is implemented with calls to `mstore8` instead of calling `mcopy`? If there was a problem with `mcopy` implementation, then I suggest we need to investigate that since `mcopy` implementation is optimized inplace memmove implementation and if there is a problem with that, we would need to find a culprit.
> > 
> > 
> > It used to break some tests, it does not do it anymore, so that change is fine
> 
> Thanks for the review! I submitted a change for `OP_MCOPY` to use `mcopy` function instead of calling `mstore8`. AFAIU, previous implementation was first copying data into intermediate buffer, then to the destination? If this is true, do we need following check, as `mcopy` implementation is doing inplace copy:
> 
> ```
>         let oldSize := mul(mload(MEM_OFFSET()),32)
>         if gt(add(oldSize,size),MAX_POSSIBLE_MEM()) {
>             revertWithGas(evmGasLeft)
>         }
> ```

Seeing that in detail, that check is actually incorrect, we still need to check that both `offset` and `destOffset` are inside the correct memory space, that can be done with:
       
```
checkMemOverflow(add(add(offset, MEM_OFFSET_INNER()), size), evmGasLeft)
checkMemOverflow(add(add(destOffset, MEM_OFFSET_INNER()), size), evmGasLeft)
```

gianbelinche commented on system-contracts/contracts/EvmInterpreterFunctions.template.yul:
```diff
@@ -1448,7 +1402,7 @@ function performCreate2(evmGas, oldSp, isStatic) -> evmGasLeft, sp, result, addr
     {
         let hashedBytecode := keccak256(add(MEM_OFFSET_INNER(), offset), size)
-        mstore8(0, 0xFF)
+        mstore(0, 0xFF000000000000000000000000000000000000000000000000000000000000)
```
Two 0s are missing here.
It has 31 bytes instead of 32.

vladimirradosavljevic replied:
Nice catch. I have fixed this.

vladimirradosavljevic commented:
> > > > @jrchatruc Do you recall a reason why `OP_MCOPY` is implemented with calls to `mstore8` instead of calling `mcopy`? If there was a problem with `mcopy` implementation, then I suggest we need to investigate that since `mcopy` implementation is optimized inplace memmove implementation and if there is a problem with that, we would need to find a culprit.
> > > 
> > > 
> > > It used to break some tests, it does not do it anymore, so that change is fine
> > 
> > 
> > Thanks for the review! I submitted a change for `OP_MCOPY` to use `mcopy` function instead of calling `mstore8`. AFAIU, previous implementation was first copying data into intermediate buffer, then to the destination? If this is true, do we need following check, as `mcopy` implementation is doing inplace copy:
> > ```
> >         let oldSize := mul(mload(MEM_OFFSET()),32)
> >         if gt(add(oldSize,size),MAX_POSSIBLE_MEM()) {
> >             revertWithGas(evmGasLeft)
> >         }
> > ```
> 
> Seeing that in detail, that check is actually incorrect, we still need to check that both `offset` and `destOffset` are inside the correct memory space, that can be done with:
> 
> ```
> checkMemOverflow(add(add(offset, MEM_OFFSET_INNER()), size), evmGasLeft)
> checkMemOverflow(add(add(destOffset, MEM_OFFSET_INNER()), size), evmGasLeft)
> ```

I have added these checks. Thanks!
Please note that I have added following to the commit message:
```
This patch is a joint work by:
Gianbelinche <gianluca.belinche@lambdaclass.com>
Vladimir Radosavljevic <vr@matterlabs.dev>
```
